### PR TITLE
杨明政ASTGNN

### PR DESCRIPTION
**Pull request by nightowl-ymz**

### CODE REVIEW

File: libcity/data/batch.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def to_tensor(self, device):
             if self.feature_name[key] == 'int':
                 self.data[key] = torch.LongTensor(np.array(self.data[key])).to(device)
             elif self.feature_name[key] == 'float':
-                self.data[key] = torch.FloatTensor(np.array(self.data[key])).to(device)
+                self.data[key] = torch.FloatTensor(np.array(self.data[key].cpu())).to(device)
```
Comment (aptx1231) [inline libcity/data/batch.py:100]: 不应该修改这个文件，这里边self.data[key]本身就是list，并不是tensor。

**Review by aptx1231 — CHANGES_REQUESTED** [review]
核心代码呢？？

File: libcity/data/batch.py [hunk 2]
```diff
@@ -128,4 +128,4 @@ def to_ndarray(self):
             else:
                 raise TypeError(
                     'Batch to_tensor, only support int, float, array of int, no_pad_float.\
-                    and you give {}'.format(self.feature_name[key]))
```
Comment (aptx1231) [inline libcity/data/batch.py:131]: 用git恢复这个文件，不要修改这个

File: libcity/data/dataset/ASTGNN.py [hunk 3]
```diff
@@ -0,0 +1,546 @@
+from libcity.data.dataset import TrafficStatePointDataset
```
Comment (aptx1231) [inline libcity/data/dataset/ASTGNN.py:1]: 这个文件名字应该是astgnn_dataset

File: libcity/data/dataset/ASTGNN.py [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+                'x': test_x_norm,
+                'target': test_target,
+                'timestamp': test_timestamp,
+            },
+            'stats': {
+                '_max': stats['_max'],
+                '_min': stats['_min'],
+            }
+        }
+        print('train x:', all_data['train']['x'].shape)
```
Comment (aptx1231) [inline libcity/data/dataset/ASTGNN.py:455]: print应该换成logger

File: libcity/data/dataset/ASTGNN.py [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+        train_target = training_set[-2]  # (B,N,T)
+        val_target = validation_set[-2]
+        test_target = testing_set[-2]
+
+        train_timestamp = training_set[-1]  # (B,1)
+        val_timestamp = validation_set[-1]
+        test_timestamp = testing_set[-1]
+
+        # max-min normalization on x
+        (stats, train_x_norm, val_x_norm, test_x_norm) = MinMaxnormalization(train_x, val_x, test_x)
```
Comment (aptx1231) [inline libcity/data/dataset/ASTGNN.py:432]: normalization 方法应该根据参数scaler得到，看看traffic_state_dataset

File: libcity/config/data/ASTGNNDatasets.json [hunk 6]
```diff
@@ -0,0 +1,13 @@
+{
+  "points_per_hour": 12,
+  "num_for_predict":12,
+  "num_of_weeks" : 0,
+  "num_of_days" : 0,
+  "num_of_hours" : 1,
+  "epochs" : 180,
+  "fine_tune_epochs" : 90,
+  "direction" : 2,
+  "adj_filename" : "raw_data/PEMSD3/PEMSD3.rel",
```
Comment (aptx1231) [inline libcity/config/data/ASTGNNDatasets.json:10]: 这三个东西本身就可以从参数dataset得到了，因为所有东西就是存在`raw_data/dataset_name/dataset_name.[rel/geo/dyna]`，不需要这三个参数。

File: libcity/data/dataset/ASTGNN.py [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+        super().__init__(config)
+        points_per_hour = int(config['points_per_hour'])
+        num_for_predict = int(config['num_for_predict'])
+        num_of_weeks = int(config['num_of_weeks'])
+        num_of_days = int(config['num_of_days'])
+        num_of_hours = int(config['num_of_hours'])
+        graph_signal_matrix_filename = config['graph_signal_matrix_filename']
+        self.id_filename = config['id_filename']
+        # data = np.load(graph_signal_matrix_filename)
+        with open(self.id_filename, 'r') as f:
```
Comment (aptx1231) [inline libcity/data/dataset/ASTGNN.py:270]: 各种读取geo/rel/dyna文件的代码在traffic_state_dataset里边都提供了，你都继承了，应该直接调用

Comment (aptx1231) [reply libcity/data/dataset/ASTGNN.py:270]: 我看这个dataset跟ASTGCN-dataset很像，其实可以参考一下，把代码简化一下

File: libcity/data/dataset/ASTGNN.py [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+                                train_x=all_data['train']['x'], train_target=all_data['train']['target'],
+                                train_timestamp=all_data['train']['timestamp'],
+                                val_x=all_data['val']['x'], val_target=all_data['val']['target'],
+                                val_timestamp=all_data['val']['timestamp'],
+                                test_x=all_data['test']['x'], test_target=all_data['test']['target'],
+                                test_timestamp=all_data['test']['timestamp'],
+                                mean=all_data['stats']['_max'], std=all_data['stats']['_min']
+                                )
+        return all_data
+    def get_data_feature(self):
```
Comment (aptx1231) [inline libcity/data/dataset/ASTGNN.py:486]: 这里边连节点个数num_nodes都不返回的话，那模型怎么用呢？必要的信息都应该返回

File: libcity/data/dataset/ASTGNN.py [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+        batch_size = int(config['batch_size'])
+        print('batch_size:', batch_size, flush=True)
+        direction = int(config['direction'])
+        # direction = 1 means: if i connected to j, adj[i,j]=1;
+        # direction = 2 means: if i connected to j, then adj[i,j]=adj[j,i]=1
+        if direction == 2:
+            self.adj_mx, distance_mx = self.get_adjacency_matrix_2direction(adj_filename)
+        if direction == 1:
+            self.adj_mx, distance_mx = self.get_adjacency_matrix(adj_filename)
+        self.train_loader, self.train_target_tensor, self.val_loader, self.val_target_tensor, self.test_loader, self.test_target_tensor, self._max, self._min = load_graphdata_normY_channel1(
```
Comment (aptx1231) [inline libcity/data/dataset/ASTGNN.py:290]: 这个位置应该写在load_data函数里边吧，并且增加cache缓存，要不然感觉现在的代码的cache没用？

File: libcity/data/dataset/ASTGNN.py [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+                                )
+        return all_data
+    def get_data_feature(self):
+        return {"adj_mx": self.adj_mx,"scaler":self.scaler}
+
+    def get_data(self):
+        train_data=[]
+        val_data=[]
+        test_data=[]
+        for batch_data in self.train_loader:
```
Comment (aptx1231) [inline libcity/data/dataset/ASTGNN.py:493]: 这个位置也比较粗糙，看看traffic_state_dataset，这里：https://github.com/LibCity/Bigscity-LibCity/blob/master/libcity/data/dataset/traffic_state_datatset.py#L942。
就是代码应该都是写好的，我觉得你这里需要改写的地方就是，_generate_input_data函数，可以看看ASTGCN-dataset，这里：https://github.com/LibCity/Bigscity-LibCity/blob/master/libcity/data/dataset/astgcn_dataset.py#L120。

File: libcity/model/traffic_flow_prediction/ASTGNN.py [hunk 11]
```diff
@@ -0,0 +1,1200 @@
...
+                nn.init.uniform_(p)
+
+    def forward(self, batch):
+        encoder_inputs = batch['encoder_inputs']
+        decoder_inputs = batch['decoder_inputs']
+        return self.model(encoder_inputs,decoder_inputs)
+
+
+    def calculate_loss(self, batch):
+        y_true = batch['y']
```
Comment (aptx1231) [inline libcity/model/traffic_flow_prediction/ASTGNN.py:1194]: 这里的数据是归一化的数据，应该进行反归一化吧。
归一化应该用框架内置的scaler：https://github.com/LibCity/Bigscity-LibCity/blob/master/libcity/utils/normalization.py#L21

File: libcity/model/traffic_flow_prediction/ASTGNN.py [hunk 12]
```diff
@@ -0,0 +1,1200 @@
...
+            nn.init.xavier_uniform_(p)
+
+    return model
+
+
+class ASTGNN(AbstractTrafficStateModel):
+    def __init__(self, config, data_feature):
+        super().__init__(config, data_feature)
+        self.Device = config.get('device', torch.device('cpu'))
+        self.num_layers = config.get('num_layers')
```
Comment (aptx1231) [inline libcity/model/traffic_flow_prediction/ASTGNN.py:1151]: 参数加上默认值。

**Review by aptx1231 — CHANGES_REQUESTED** [review]
主要还是dataset的问题，建议参考astgcn-dataset，把代码简化一下。
冗余的代码应该借用父类的函数。
